### ISSUE TITLE
Initial simulations - Happy Flow

### ISSUE THREAD

**Issue by ronentk** [orbs-network/consensus-playground#4]
Vary `M`,`N`,`k`, `blockSize` and plot:
- Confirmation time/Latency - time until block propagates throughout network
- Throughput - `(blockSize/txSize) / confirmationTime` 

Need to model bandwidth and link latency (latency matrix)

**Comment by ronentk** [comment]
@OdedWx @gadcl  @idandash @avilanthe1 @orbs-network/researchers 

Some results for initial happy flow simulations with following parameters:

```
const TOTAL_NODE_FACTORS = [1]; // ratio between m an d n
const COMMITTEE_SIZES = [4, 8, 16, 32, 64];
const SHARING_THRESHOLD_FACTORS = [1]; // ratio between k and f+1
const ETX_SIZES_BYTES = [250];
const NUM_ETXS_PER_BLOCK =  [128, 256, 512, 1024, 2048, 4096, 8192, 16384];
const NODE_BW_BITS_SEC = [1000000000, 500000000, 100000000];
const MAX_SIMULATION_TIMESTAMP_MS = 100000;
```
[results](https://github.com/orbs-network/private-consensus-playground/files/1825455/bf-1-1.html.zip)

![thrput-latency](https://user-images.githubusercontent.com/1759025/37600810-0df88e6a-2b91-11e8-9a79-01cdc59d0bc5.png)
![bw1](https://user-images.githubusercontent.com/1759025/37602920-1c9bc91e-2b96-11e8-8e97-29251f604c23.png)

**Assumptions**
- All nodes are committee members in this set of simulations
- Happy flow, no faults
- Network latency varies between 10-370 ms per message, based on site recommended by @OdedWx . Added to this are bandwidth bottlenecks on the transmission and reception ends. Message sizes are detailed [here](https://docs.google.com/spreadsheets/d/1CI87VXOd4aRjpAP6TM7vGaGxJjsX5CuSso6vp3YJrlI/edit#gid=1987181903)

**Remarks**
- Low bandwidth scenarios are causing trouble even with happy flow for all but the smallest committees. Looks like this is partly due to timeout duration, but we'll need to look into it further.
- Proposal timeout set for 2 seconds. Meaning that at the beginning of a term, committee nodes set the timer and if the block isn't closed after 2 seconds, they start the new view procedure. This can be optimised (for example by resetting timer after each PBFT stage).
- Checked effect of ratio between `k/f`. Increasing it to `1.5` didn't seem to significantly affect results. (graphs not included)

## Next steps (feel free to add your own):
**Simulation**
- [x] Vary timeout duration to see effect on low bandwidth scenarios.
- [ ] Increase block size until we see saturation for cases of small committees/large BW
- [ ] Averaging simulation over multiple seeds to smooth results
- [x] Check effects of varying `m/n` ratio.
- [x] Check effects of varying `k/(f+1)` ratio.
- [ ] Add partitioned networks (different clouds)
 
**Analysis**
- [x] Low bandwidth scenarios- and analyse logs to understand sources of problems.
- [ ] Understand bottlenecks - contribution of each stage of algorithm (consensus, decryption)

**Infrastructure / Code**
- [ ] Multithreading to run simulations faster (#6 )
- [ ] SSH support to enable remote running on Orbs research machine
- [x] Add timeout duration as hyperparameter
- [x] Ping before sending blocks/shares to avoid sending duplicates - WIP by @gadcl

**Comment by ronentk** [comment]
Regarding bottlenecks:
First [obvious] bottleneck is the transmission of the leader's pre-prepare message:
Assuming committee size `m`, block size `B` Mbytes and bandwidth capacity of `c` bytes/ ms
The time it takes (in ms) to send all copies of the pre-prepare message is currently on the order of `TP = m * ( B / c )` since we're doing a simple multicast by the leader.

For `B=2.5*10^6` bytes (for 5000 etx / block, 250 bytes / etx ) and `c=12500` (for node bandwidth of 100 Mbits / sec) it takes around `200 * m` ms per block (disregarding network delay). This obviously doesn't scale for large committee sizes, for `m=20` we get a 4 sec period before all members get the block.

Note that this scales linearly with `c`, so for 1Gb bandwidths the delay becomes manageable which is also what we saw in the simulations. However, reliable 1Gb bandwidths probably don't exist in the wild.

How do standard PBFT implementations handle this? @gadcl

**Comment by ronentk** [comment]
Regarding the BW assumptions, after checking with @OdedWx it seems we can assume for now that nodes will have 1Gb / sec bandwidths, so a bandwidth of 100 Mbits/ sec could be considered Byzantine. In this case we can achieve a rate of thousands of txs per second at block times of < 5 sec for up to `m=32`. Increasing the timeout indeed helped stabilize the network for high-load regimes (large committees/large block sizes).

Experiments in commit 6adb97a, notebook to explore results [here](https://github.com/orbs-network/private-consensus-playground/blob/6adb97a3755735d888b515e698d5609e7387376c/simulations/happy-flow-bw-1-nb.ipynb)

![bw1g](https://user-images.githubusercontent.com/1759025/37906683-92886b4e-310b-11e8-88bd-db4e3284d3de.png)

For higher bandwidths, throughput can rise to 10s of thousands of txs:
![bw_1](https://user-images.githubusercontent.com/1759025/37906750-c447e100-310b-11e8-8de8-1aa087bcb80f.png)

**Comment by ronentk** [comment]
Discussed results with Oded, some additional points to consider:
- Transaction gossiping effect on node bandwidth (maybe find some upper bound percentage of node bandwidth since we aren't currently simulating transactions)
- Possible optimizations for leader pre-prepare message transmission

**Comment by ronentk** [comment]
[Results](https://github.com/orbs-network/private-consensus-playground/blob/d9b7c14c46a5dcc63fad33bbee4e7992d0501591/simulations/happy-flow-bw-1-nb.ipynb) for varying `n/m` ratio

TL;DR - 
- for a given `m`, doubling `n` from `n=m` to `n=2*m` leads to a slight decrease in throughput, this effect weakens as `m` grows.
- for a given `m`, increasing `k` from `k=f+1` to `k=2*f+1` leads to a slight decrease in throughput, this effect weakens as `m` grows.
